crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder/automatic_shield_builder.rs

```diff
@@ -0,0 +1,916 @@
+
+    /// While we assign factors to Recovery and Confirmation roles, we remove
+    /// them from this set.
+    remaining_available_factors: IndexSet<FactorSource>,
+
+    /// The factors assigned to each role, including the factors originally
+    /// set for the primary role.
+    proto_matrix: ProtoMatrix,
+}
+
+impl SecurityShieldBuilder {
+    /// Assigns the factors to the Recovery and Confirmation roles according to the heuristic
+    /// laid out in ["Automatic Security Shield Construction" document][doc].
+    ///
+    /// The `all_factors_in_profile` should contain all factors that are available for the user,
+    /// i.e. from Profile, and SHOULD contain the factors that the user has preselected for the
+    /// Primary role.
+    ///
+    /// # Throws
+    /// Throws if the primary role is invalid.
+    /// Throws if the primary override factors are not empty.
+    /// Throws if the primary factors are not in the profile.
+    /// Throws if the prerequisites are not met.
+    /// Throws if the shield is invalid after auto assignment - which should not happen, can be considered programmer error!
+    ///
+    /// [doc]: https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3758063620/MFA+Rules+for+Factors+and+Security+Shields#Automatic-Security-Shield-Construction
+    pub fn auto_assign_factors_to_recovery_and_confirmation_based_on_primary(
+        &self,
+        all_factors_in_profile: IndexSet<FactorSource>,
+    ) -> Result<AutoBuildOutcomeForTesting>
+/* Feel free to replace `AutoBuildOutcomeForTesting` return type if you need anything else, I had Unit, so might as well make testing easier by returning this type. */
+    {
+        if let Some(invalid_reason) = self.validate_primary_role() {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: format!(
+                    "Primary role is not valid: {:?}",
+                    invalid_reason
+                ),
+            });
+        }
+
+        if !self.get_primary_override_factors().is_empty() {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: "Primary override factors not allowed when preselecting factors for Recovery and Confirmation".to_string(),
+            });
+        }
+
+        let primary_factors = self
+            .get_primary_threshold_factors()
+            .into_iter()
+            .collect::<IndexSet<_>>();
+
+        if primary_factors
+            .intersection(
+                &all_factors_in_profile
+                    .iter()
+                    .map(|f| f.id())
+                    .collect::<IndexSet<_>>(),
+            )
+            .cloned()
+            .collect::<IndexSet<FactorSourceID>>()
+            != primary_factors
+        {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: "Primary factors not in profile".to_string(),
+            });
+        }
```

Comment:
I guess this is good to have, though we should be able to build the primary with factors not in profile in the first place.

Reply:
I like asserts, as you know :) always helpful!